src/code/sys_math3d.c

```diff
@@ -0,0 +1,1691 @@
+        osSyncPrintf("cross = pos を返します。\n");                             // "Returns cross = pos.
+        osSyncPrintf(VT_RST);
+        Math_Vec3f_Copy(ret, pos);
+    }
+    temp_f0 =
+        (((pos->x - line->a.x) * line->b.x) + ((pos->y - line->a.y) * line->b.y) + ((pos->z - line->a.z) * line->b.z)) /
+        temp_ret;
+    ret->x = (line->b.x * temp_f0) + line->a.x;
+    ret->y = (line->b.y * temp_f0) + line->a.y;
+    ret->z = (line->b.z * temp_f0) + line->a.z;
+}
+
+void func_800CACAC(f32 arg0, f32 arg1, f32 arg2, f32 arg3, f32 arg4, f32 arg5, f32 arg6, f32* arg7, f32* arg8) {
+    *arg7 = ((arg1 * arg6) - (arg3 * arg5)) / arg4;
+    *arg8 = ((arg2 * arg5) - (arg0 * arg6)) / arg4;
+}
+
+s32 func_800CAD08(f32 arg0, f32 arg1, f32 arg2, f32 arg3, f32 arg4, f32 arg5, f32 arg6, f32 arg7, Linef* arg8) {
+    char pad[4];
+    Vec3f sp60;
+    Vec3f sp54;
+    f32 ax;
+    f32 ay;
+    f32 az;
+
+    VEC3F(sp60, arg0, arg1, arg2);
+    VEC3F(sp54, arg4, arg5, arg6);
+
+    Math3D_Vec3f_Cross(&sp60, &sp54, &arg8->b);
+
+    if (fabsf(arg8->b.x) < 0.008f && fabsf(arg8->b.y) < 0.008f && fabsf(arg8->b.z) < 0.008f) {
+        return 0;
+    }
+
+    ax = fabsf(arg8->b.x);
+    ay = fabsf(arg8->b.y);
+    az = fabsf(arg8->b.z);
+
+    if ((ay <= ax) && (az <= ax)) {
+        func_800CACAC(arg1, arg2, arg5, arg6, arg8->b.x, arg3, arg7, &arg8->a.y, &arg8->a.z);
+        arg8->a.x = 0.0f;
+    } else if ((ax <= ay) && (az <= ay)) {
+        func_800CACAC(arg2, arg0, arg6, arg4, arg8->b.y, arg3, arg7, &arg8->a.z, &arg8->a.x);
+        arg8->a.y = 0.0f;
+    } else {
+        func_800CACAC(arg0, arg1, arg4, arg5, arg8->b.z, arg3, arg7, &arg8->a.x, &arg8->a.y);
+        arg8->a.z = 0.0f;
+    }
+    return 1;
+}
+
+s32 func_800CAEE8(f32 arg0, f32 arg1, f32 arg2, f32 arg3, f32 arg4, f32 arg5, f32 arg6, f32 arg7, Vec3f* arg8,
+                  Vec3f* arg9) {
+    static Linef D_8016A5D0;
+    if (func_800CAD08(arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7, &D_8016A5D0) == 0) {
+        return 0;
+    }
+    Math3D_LineVsPos(&D_8016A5D0, arg8, arg9);
+    return 1;
+}
+
+void func_800CAF5C(Vec3f* arg0, Vec3f* arg1, f32 arg2, Vec3f* arg3) {
+    arg3->x = (arg1->x * arg2) + arg0->x;
+    arg3->y = (arg1->y * arg2) + arg0->y;
+    arg3->z = (arg1->z * arg2) + arg0->z;
+}
+
+void func_800CAFA0(Vec3f* v0, Vec3f* v1, f32 arg2, Vec3f* ret) {
+    Vec3f diff;
+
+    Math_Vec3f_Diff(v1, v0, &diff);
+    func_800CAF5C(v0, &diff, arg2, ret);
+}
+
+f32 Math3D_DotProduct(Vec3f* vec1, Vec3f* vec2) {
+    f32 ret;
+
+    Math3D_CalcDotProduct(vec1, vec2, &ret);
+    return ret;
+}
+
+s32 Math3D_CalcDotProduct(Vec3f* vec1, Vec3f* vec2, f32* dst) {
+    f32 magProduct;
+
+    magProduct = Math3D_Vec3fMagnitude(vec1) * Math3D_Vec3fMagnitude(vec2);
+    if (fabsf(magProduct) < 0.008f) {
+        *dst = 0.0f;
+        return 1;
+    }
+    *dst = ((vec1->x * vec2->x) + (vec1->y * vec2->y) + (vec1->z * vec2->z)) / magProduct;
+    return 0;
+}
```

Comment:
This is specifically the dot product divided by the product of the vector magnitudes rather than the dot product itself.
(Which is also the cosine of the angle between the two vectors it was calculated from)